### ISSUE TITLE
Build failed in phase LLVM Optimiser (Exit code: -11)

### ISSUE THREAD

**Issue by 2bdkid** [harendra-kumar/unicode-transforms#10]
I'm installing on a Raspberry Pi 2 with the prebuilt GHC package. I don't know if this error is ARM specific but this is the log.

cabal: Entering directory '/tmp/cabal-tmp-7667/unicode-transforms-0.2.0'
Configuring unicode-transforms-0.2.0...
Building unicode-transforms-0.2.0...
Preprocessing library unicode-transforms-0.2.0...
[ 1 of 12] Compiling Data.Unicode.Properties.DecompositionsK ( Data/Unicode/Properties/DecompositionsK.hs, dist/build/Data/Unicode/Properties/DecompositionsK.o )
[ 2 of 12] Compiling Data.Unicode.Properties.Decompositions ( Data/Unicode/Properties/Decompositions.hs, dist/build/Data/Unicode/Properties/Decompositions.o )
[ 3 of 12] Compiling Data.Unicode.Properties.DecomposableK ( Data/Unicode/Properties/DecomposableK.hs, dist/build/Data/Unicode/Properties/DecomposableK.o )
\`opt' failed in phase `LLVM Optimiser'. (Exit code: -11)
cabal: Leaving directory '/tmp/cabal-tmp-7667/unicode-transforms-0.2.0'

**Comment by harendra-kumar** [comment]
Can you provide `ghc --version` and `opt --version` please?

**Comment by 2bdkid** [comment]
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1

$ opt --version
LLVM (http://llvm.org/):
  LLVM version 3.7.1
  Optimized build.
  Default target: armv7l-unknown-linux-gnueabihf
  Host CPU: (unknown)

I'm on Ubuntu Mate

**Comment by harendra-kumar** [comment]
I just compiled it with llvm (``stack build --flag unicode-transforms:has-llvm``) on x86 machine:

```
$ stack ghc -- --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1

$ opt --version
LLVM (http://llvm.org/):
  LLVM version 3.7.1
  
  Optimized build.
  Built Oct  7 2016 (06:55:25).
  Default target: x86_64-pc-linux-gnu
  Host CPU: haswell
```

It looks like an ARM specific issue. Can you raise an issue on ARM GHC instead?

**Comment by 2bdkid** [comment]
GHC's bug tracker says it's an out of memory exit code. Do you know if it's possible to disable optimization?

**Comment by 2bdkid** [comment]
The RPI itself has 1gb of ram. I added a swap file (~7gb) and it fixed issues with other packages running out of memory, but unicode-transforms still fails

**Comment by harendra-kumar** [comment]
I split the big file into two parts. Can you try compiling the latest code from the github repo?

**Comment by 2bdkid** [comment]
Still error -11, even with the swap file

**Comment by harendra-kumar** [comment]
Error on which file? Can you paste the error here?

**Comment by harendra-kumar** [comment]
Ah, I split the wrong file. Your original problem was with DecomposableK and not DecompositionsK.

**Comment by harendra-kumar** [comment]
Can you please try now? I reduced the size of the input list to bitarray by half. Hopefully this will fix the llvm optimizer issue.

**Comment by brent80** [comment]
While building package unicode-transforms-0.3.0 using:
      /home/lumberjack/.stack/setup-exe-cache/arm-linux/Cabal-simple_mPHDZzAJ_1.24.2.0_ghc-8.0.2 --builddir=.stack-work/dist/arm-linux/Cabal-1.24.2.0 build --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/lumberjack/Work/stack/.stack-work/logs/unicode-transforms-0.3.0.log

    Configuring unicode-transforms-0.3.0...
    Building unicode-transforms-0.3.0...
    Preprocessing library unicode-transforms-0.3.0...
    [ 1 of 14] Compiling Data.Unicode.Types ( Data/Unicode/Types.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Types.o )
    [ 2 of 14] Compiling Data.Unicode.Properties.DecompositionsK2 ( Data/Unicode/Properties/DecompositionsK2.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/DecompositionsK2.o )
    [ 3 of 14] Compiling Data.Unicode.Properties.DecompositionsK ( Data/Unicode/Properties/DecompositionsK.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/DecompositionsK.o )
    [ 4 of 14] Compiling Data.Unicode.Properties.Decompositions ( Data/Unicode/Properties/Decompositions.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/Decompositions.o )
    [ 5 of 14] Compiling Data.Unicode.Properties.DecomposeHangul ( Data/Unicode/Properties/DecomposeHangul.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/DecomposeHangul.o )
    [ 6 of 14] Compiling Data.Unicode.Properties.Compositions ( Data/Unicode/Properties/Compositions.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/Compositions.o )
    [ 7 of 14] Compiling Data.Unicode.Properties.BitArray ( Data/Unicode/Properties/BitArray.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/BitArray.o )
    [ 8 of 14] Compiling Data.Unicode.Properties.CombiningClass ( Data/Unicode/Properties/CombiningClass.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/CombiningClass.o )
    [ 9 of 14] Compiling Data.Unicode.Properties.Decomposable ( Data/Unicode/Properties/Decomposable.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/Decomposable.o )
    [10 of 14] Compiling Data.Unicode.Properties.DecomposableK ( Data/Unicode/Properties/DecomposableK.hs, .stack-work/dist/arm-linux/Cabal-1.24.2.0/build/Data/Unicode/Properties/DecomposableK.o )
    qemu: uncaught target signal 11 (Segmentation fault) - core dumped
    `opt' failed in phase `LLVM Optimiser'. (Exit code: -11)

**Comment by harendra-kumar** [comment]
Can you try compiling just that file i.e. DecomposableK and see if it compiles? You can use `stack ghc Data.Unicode.Properties.DecomposableK` to compile it.

I have never compiled the package on arm. Can you help me in compiling it on arm, what all do I need? I definitely do not have an arm hardware other than a phone.

It is most likely an llvm problem but don't know what factors cause llvm to blow up, and if there is anything that we can do in our source to avoid the problem.

**Comment by brent80** [comment]
Cloned your library and ran that command, got this
Data/Unicode/Properties/BitArray.hs:15:8:
    Could not find module ‘Data.BitArray.ST’
    Perhaps you meant
      Data.Array.ST (from array-0.5.1.0@array_67iodizgJQIIxYVTp4emlA)
    Use -v to see a list of the files searched for.

Data/Unicode/Properties/BitArray.hs:17:8:
    Could not find module ‘Data.BitArray’
    Perhaps you meant
      Data.Array (from array-0.5.1.0@array_67iodizgJQIIxYVTp4emlA)
    Use -v to see a list of the files searched for.

**Comment by brent80** [comment]
After switching to lts-8.5 I got this
[1 of 2] Compiling Data.Unicode.Properties.BitArray ( Data/Unicode/Properties/BitArray.hs, Data/Unicode/Properties/BitArray.o )
[2 of 2] Compiling Data.Unicode.Properties.DecomposableK ( Data/Unicode/Properties/DecomposableK.hs, Data/Unicode/Properties/DecomposableK.o )
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
`opt' failed in phase `LLVM Optimiser'. (Exit code: -11)

**Comment by harendra-kumar** [comment]
Can you watch the memory consumption while this file is being compiled, using `top` or something equivalent?

**Comment by brent80** [comment]
It has to be some kind of instant memory spike.....because I can't see it go up
it goes from like 0.2% to crash

**Comment by harendra-kumar** [comment]
What is the llvm version you are using? Not sure if ghc will work with any llvm version, if it does we can try a newer llvm.

**Comment by brent80** [comment]
3.7.1 is the newest we can use....

**Comment by brent80** [comment]
i'm trying to build the new stack, and it requires this library

**Comment by brent80** [comment]
It can be quite difficult, but you can setup qemu arm

**Comment by harendra-kumar** [comment]
Ah, ok, let me try asking @borsboom how the original stack binary was built in the first place. Why it did not fail then and is failing for you?

**Comment by brent80** [comment]
In a nutshell you take a raspberry pi image you mount it
and then chroot into the image while using qemu as a 
raspberry pi emulator to be able to run the files in the image
as armv7.  Once inside the image you can do about anything
you want.  You can install ghc and llvm and compile code. This article is where I found this info.
https://lionfacelemonface.wordpress.com/2015/04/18/raspberry-pi-build-environment-in-no-time-at-all/

These were the commands I used.

    sudo apt-get install kpartx pv  
    wget http://static.proot.me/proot-x86_64                                                
    chmod +x proot-x86_64                                                           
    sudo mv proot-x86_64 /usr/local/bin/proot

downlaod qemu
http://wiki.qemu.org/Download

     sudo apt-get install build-essential pkg-config libzip-dev libglib2.0-dev zlib1g-dev autoconf libtool -y
     cd ~/Downloads && tar xf qemu-2.6.0.tar.bz2
     cd qemu-2.6.0/
     ./configure
     make -j4
     sudo make install

sudo kpartx -a -v 2015-02-16-raspbian-wheezy.img

ls /dev/mapper shows you the created loops

mkdir raspbian-boot raspbian-root make mount directories

sudo mount /dev/mapper/loop0p1 raspbian-boot

sudo mount /dev/mapper/loop0p2 raspbian-root

you can set the stack size like this in next command -q "qemu-arm -s <bytes>"

sudo proot -q qemu-arm -S raspbian-root -b raspbian-boot:/boot /bin/bash

Once you are in the image you have to add the path to your root user 

export PATH=$PATH:/usr/sbin:/sbin:/bin/:/usr/local/sbin/:/home/pi/.cabal/bin:/home/pi/.yesod/.cabal-
sandbox/bin:/home/pi/.ghc/bin:/home/pi/Work/node-v5.0.0-linux-armv7l/bin

or you can su into a user on that image example su pi

Now you can install ghc for pi and install llvm 3.5.2 just move opt and llc to /usr/bin

Make sure you are using the gold linker

sudo ln -sf /usr/bin/ld /usr/bin/ld.gold

Now compile your raspberry pi code and then you can 

exit

sync

sudo umount raspbian-root raspbian-boot

sudo kpartx -d 2015-02-16-raspbian-wheezy.img

Add it is back to image form which can be transfered back to

sd card and ran on pi.

**Comment by brent80** [comment]
He was able to build stack 1.4 for arm?

**Comment by brent80** [comment]
that guide is a bit dated, you can get newer version of qemu

**Comment by brent80** [comment]
It couldn't be a 32bit problem, like something that doesn't happen on a 64bit? just know I have ran into that sometimes in the past, but usually doesn't cause compile issues, just runtime

**Comment by harendra-kumar** [comment]
stack releases an arm binary distribution, I can see it on the release page up to release 1.3.2 (https://github.com/commercialhaskell/stack/releases/download/v1.3.2/stack-1.3.2-linux-arm.tar.gz) . unicode-transforms does not change much, if it compiled for earlier versions of stack it should compile for 1.4 as well. @borsboom might have a plan to create an arm binary for 1.4 as well.

**Comment by brent80** [comment]
Yeah, I think unicode-transforms might have gotten added to this version
not sure on that though.  I have built the previous versions though.

**Comment by harendra-kumar** [comment]
unicode-transforms has been there for quite some time, it was definitely there in 1.3.2. That's why I am wondering why is it failing now. It may be something specific to your setup.

**Comment by brent80** [comment]
Ok, well thanks for trying to look into for me.

**Comment by brent80** [comment]
I tried building all of your versions from 2.0 up I think, they all failed for me

**Comment by harendra-kumar** [comment]
Its a segmentation fault, we will need a core dump to investigate what's going wrong with llvm opt. I will try creating an arm setup sometime and see if I can reproduce this. Meanwhile, stack team might build an arm version of stack 1.4 for you.

**Comment by brent80** [comment]
that would great, thank you

**Comment by brent80** [comment]
I have an update.  I switched back to ghc-7.10.3 and used llvm 3.5 binaries (opt and llc) and it built.  I'm not sure what the issue is, but I think I will try build stack with this setup now.

**Comment by harendra-kumar** [comment]
https://github.com/commercialhaskell/stack/issues/2103 is related to getting stack to officially work on ARM. It may have some relevant info.

**Comment by borsboom** [comment]
FYI: I'm running into this same exact issue, following the same process to build a stack-1.5.0 arm binary as I did for the stack-1.3.2 binary.

**Comment by brent80** [comment]
@borsboom @harendra-kumar 
I submitted a pr to fix the llvm build

**Comment by borsboom** [comment]
Cool!  Still a change in behaviour in newer GHC/LLVM that indicates a bug *somewhere* but this workaround will be most helpful for future Stack builds.

**Comment by harendra-kumar** [comment]
Awesome! Thanks @brent80! I will take a look at the PR and merge it.

**Comment by harendra-kumar** [comment]
Fixed by be05072.
